File: Source/Relativity.Import.Client.NUnit/MockBulkImageFileImporter.cs
// -----------------------------------------------------------------------------------------------------
// <copyright file="MockBulkImageFileImporter.cs" company="Relativity ODA LLC">
//   © Relativity All Rights Reserved.
// </copyright>
// -----------------------------------------------------------------------------------------------------

namespace Relativity.Import.Client.NUnit
{
	using System;
	using System.IO;
	using System.Text;
	using System.Threading;

	using kCura.WinEDDS;

	using Relativity.Import.Export.Io;
	using Relativity.Import.Export.Process;
	using Relativity.Logging;

	using ExecutionSource = Relativity.Import.Export.Services.ExecutionSource;

	/// <summary>
	/// Represents a mock class object for <see cref="BulkImageFileImporter"/>.
	/// </summary>
	public class MockBulkImageFileImporter : BulkImageFileImporter
	{
		public MockBulkImageFileImporter(
			ImageLoadFile args,
			ProcessContext context,
			IIoReporter reporter,
			ILog logger,
			Guid processID,
			bool doRetryLogic,
			kCura.WinEDDS.Service.IBulkImportManager manager,
			kCura.WinEDDS.Api.IImageReader reader,
			CancellationTokenSource tokenSource,
			ExecutionSource executionSource)
			: base(
				0,
				args,
				context,
				reporter,
				logger,
				processID,
				doRetryLogic,
				false,
				tokenSource,
				executionSource)
		{
			this._bulkImportManager = manager;
			this._imageReader = reader;

			this.ImportBatchVolume = 1000000;
			this.OutputFromStringWriter = new StringBuilder();
		}

		public int BatchSize
		{
			get => this.ImportBatchSize;
			set => this.ImportBatchSize = value;
		}

		public string InputForStringWriter
		{
			get;
			set;
		}

		public int MinimumBatch
		{
			get => this.MinimumBatchSize;
			set => this.MinimumBatchSize = value;
		}

		public StringBuilder OutputFromStringWriter
		{
			get;
		}

		public int PauseCalled
		{
			get;
			private set;
		}

		protected override bool BatchResizeEnabled => true;

		protected override int NumberOfRetries => 3;

		protected override int WaitTimeBetweenRetryAttempts => 0;

		public void SetImportBatchSize(int size)
		{
			this.ImportBatchSize = size;
		}

		public void MockLowerBatchSizeAndRetry(int numberOfRecords)
		{
			this.LowerBatchSizeAndRetry(string.Empty, string.Empty, numberOfRecords);
		}

		public kCura.EDDS.WebAPI.BulkImportManagerBase.MassImportResults TryBulkImport(kCura.EDDS.WebAPI.BulkImportManagerBase.OverwriteType overwrite)
		{
			const bool UseBulk = false;
			return this.RunBulkImport(overwrite, UseBulk);
		}

		protected override TextWriter CreateStreamWriter(string tmpLocation)
		{
			return new StringWriter(this.OutputFromStringWriter);
		}

		protected override TextReader CreateStreamReader(string outputPath)
		{
			return new StringReader(this.InputForStringWriter);
		}

		protected override void InitializeDTOs(ImageLoadFile args)
		{
			// Do NOT call the base class!
			this._keyFieldDto = new kCura.EDDS.WebAPI.FieldManagerBase.Field();
			this._keyFieldDto.ArtifactID = 0;
		}

		protected override void InitializeUploaders(ImageLoadFile args)
		{
			// Do NOT call the base class!
		}

		protected override void RaiseWarningAndPause(Exception exception, int timeoutSeconds, int retryCount, int totalRetryCount)
		{
			// Do NOT call the base class!
			this.PauseCalled++;
		}

		protected override int DoLogicAndPushImageBatch(
			int totalRecords,
			int recordsProcessed,
			string bulkLocation,
			string dataGridLocation,
			ref long charactersSuccessfullyProcessed,
			int i,
			long charactersProcessed)
		{
			return 100;
		}
	}
}

File: Source/Relativity.DataExchange.Import.Legacy.NUnit/MockBulkImageFileImporter.vb
' -----------------------------------------------------------------------------------------------------
' <copyright file="MockBulkImageFileImporter.cs" company="Relativity ODA LLC">
'   © Relativity All Rights Reserved.
' </copyright>
' -----------------------------------------------------------------------------------------------------

Imports System.Threading

Imports kCura.EDDS.WebAPI.BulkImportManagerBase
Imports kCura.EDDS.WebAPI.DocumentManagerBase
Imports kCura.WinEDDS
Imports Relativity.DataExchange.Io
Imports Relativity.DataExchange.Process

Imports Relativity.Logging

Namespace Relativity.DataExchange.Import.NUnit

	''' <summary>
	''' Inheriting the BulkLoadFileImporter to override its BulkImportManager property (the part we want to throw timeout exceptions)
	''' </summary>
	Public Class MockBulkImageFileImporter
		Inherits kCura.WinEDDS.BulkImageFileImporter

		Private Property WillThrowException As Boolean
		Public Property BatchSize As Int32
			Get
				Return Me.ImportBatchSize
			End Get
			Set(ByVal value As Int32)
				ImportBatchSize = value
			End Set
		End Property

		Protected Overrides ReadOnly Property BatchResizeEnabled As Boolean
			Get
				Return True
			End Get
		End Property

		Public PauseCalled As Int32 = 0

		Public Sub New(ByVal args As ImageLoadFile, ByVal processController As ProcessContext, ByVal ioReporterInstance As IIoReporter, ByVal logger As ILog, ByVal processID As Guid, ByVal doRetryLogic As Boolean, ByVal throwsException As Boolean, ByVal bulkManager As kCura.WinEDDS.Service.BulkImportManager, ByVal tokenSource As CancellationTokenSource)
			MyBase.New(0, args, processController, ioReporterInstance, logger, processID, doRetryLogic, tokenSource, AddressOf GetCorrelationIdFunc)
			Me.WillThrowException = throwsException
			_bulkImportManager = bulkManager
			Me.ImportBatchSize = 500
			Me.ImportBatchVolume = 1000000
		End Sub

		Public Function TryBulkImport(ByVal overwrite As kCura.EDDS.WebAPI.BulkImportManagerBase.OverwriteType) As MassImportResults
			Return RunBulkImport(overwrite, False)
		End Function

		Protected Overrides Sub InitializeDTOs(ByVal args As ImageLoadFile)
			_keyFieldDto = New kCura.EDDS.WebAPI.FieldManagerBase.Field
			_keyFieldDto.ArtifactID = 0
		End Sub

		Protected Overrides Sub InitializeUploaders(ByVal args As ImageLoadFile)
			'do nothing
		End Sub

		Protected Overrides Sub RaiseWarningAndPause(ByVal ex As System.Exception, ByVal timeoutSeconds As Integer)
			PauseCalled += 1
		End Sub

		Public Property MinimumBatch As Integer
			Get
				Return MyBase.MinimumBatchSize
			End Get
			Set(ByVal value As Integer)
				MinimumBatchSize = value
			End Set
		End Property

		Protected Overrides ReadOnly Property NumberOfRetries As Int32
			Get
				Return 3
			End Get
		End Property

		Private Shared Function GetCorrelationIdFunc() As String
			Return NameOf(MockBulkImageFileImporter)
		End Function
	End Class
End Namespace